features/netsocket/emac-drivers/TARGET_NUVOTON_EMAC/numaker_emac.cpp

```diff
@@ -0,0 +1,400 @@
+            /* Copy data to Tx buffer*/
+            memcpy(static_cast<uint8_t *>(buffer) + bufferoffset, static_cast<uint8_t *>(memory_manager->get_ptr(q)) + payloadoffset, (PACKET_BUFFER_SIZE - bufferoffset));
+
+            /* Point to next descriptor */
+            numaker_eth_trigger_tx(PACKET_BUFFER_SIZE, NULL);
+            buffer = numaker_eth_get_tx_buf();
+            if( buffer == NULL ) goto error;
+            
+            byteslefttocopy = byteslefttocopy - (PACKET_BUFFER_SIZE - bufferoffset);
+            payloadoffset = payloadoffset + (PACKET_BUFFER_SIZE - bufferoffset);
+            framelength = framelength + (PACKET_BUFFER_SIZE - bufferoffset);
+            bufferoffset = 0;
+        }
+        
+        /* Copy the remaining bytes */
+        memcpy(static_cast<uint8_t *>(buffer) + bufferoffset, static_cast<uint8_t *>(memory_manager->get_ptr(q)) + payloadoffset, byteslefttocopy);
+        bufferoffset = bufferoffset + byteslefttocopy;
+        framelength = framelength + byteslefttocopy;
+    }
+    
+    /* Prepare transmit descriptors to give to DMA */
+    numaker_eth_trigger_tx(framelength, NULL);
+
+    result = true;
+
+error:
+
+    memory_manager->free(buf);
+
+    /* Restore access */
+    TXLockMutex.unlock();
+
+    return result;
+}
+
+
+void NUMAKER_EMAC::phy_task()
+{
+    
+    // Get current status
+    int state;
+    state = numaker_eth_link_ok();
+
+    
+    if ((state & PHY_LINKED_STATE) && !(phy_state & PHY_LINKED_STATE)) {
+        NU_DEBUGF(("Link Up\r\n"));
+        if (emac_link_state_cb) emac_link_state_cb(true);
+    } else if (!(state & PHY_LINKED_STATE) && (phy_state & PHY_LINKED_STATE)) {
+        NU_DEBUGF(("Link Down\r\n"));
+        if (emac_link_state_cb) emac_link_state_cb(false);
+    }
+    phy_state = state;        
+
+}
+
+bool NUMAKER_EMAC::power_up()
+{
+  /* Initialize the hardware */
+  if (!low_level_init_successful())
+    return false;
+
+  /* Worker thread */
+  thread = create_new_thread("numaker_emac_thread", &NUMAKER_EMAC::thread_function, this, THREAD_STACKSIZE, THREAD_PRIORITY, &thread_cb);
+
+  /* PHY monitoring task */
+  phy_state = PHY_UNLINKED_STATE;
+
+  phy_task_handle = mbed::mbed_event_queue()->call_every(PHY_TASK_PERIOD_MS, mbed::callback(this, &NUMAKER_EMAC::phy_task));
+
+  /* Allow the PHY task to detect the initial link state and set up the proper flags */
+  osDelay(10);
+  numaker_eth_enable_interrupts();
+  return true;
+}
+
+
+uint32_t NUMAKER_EMAC::get_mtu_size() const
+{
+  return NU_ETH_MTU_SIZE;
+}
+
+uint32_t NUMAKER_EMAC::get_align_preference() const
+{
+  return NU_BUFF_ALIGNMENT;
+}
+
+void NUMAKER_EMAC::get_ifname(char *name, uint8_t size) const
+{
+  memcpy(name, NU_ETH_IF_NAME, (size < sizeof(NU_ETH_IF_NAME)) ? size : sizeof(NU_ETH_IF_NAME));
+}
+
+uint8_t NUMAKER_EMAC::get_hwaddr_size() const
+{
+    return NU_HWADDR_SIZE;
+}
+
+bool NUMAKER_EMAC::get_hwaddr(uint8_t *addr) const
+{
+    mbed_mac_address((char *)addr);
```

Comment:
If you can implement "set_hwaddr", and don't have a firm idea of a specific address for this interface, the "get_hwaddr" function should return false to say "I don't have my own ID".

The stack is expected to come up with a MAC itself (probably by calling mbed_mac_address) if the interface has no ID chip of its own.

This will do for now though